app/models/patient.rb

```diff
@@ -774,8 +800,8 @@ def self.from_fhir(patient)
       first_name: patient&.name&.first&.given&.first,
       middle_name: patient&.name&.first&.given&.second,
       last_name: patient&.name&.first&.family,
-      primary_telephone: Phonelib.parse(patient&.telecom&.select { |t| t&.system == 'phone' }&.first&.value, 'US').full_e164,
-      secondary_telephone: Phonelib.parse(patient&.telecom&.select { |t| t&.system == 'phone' }&.second&.value, 'US').full_e164,
+      primary_telephone: PatientHelper.from_fhir_phone_number(patient&.telecom&.select { |t| t&.system == 'phone' }&.first&.value),
+      secondary_telephone: PatientHelper.from_fhir_phone_number(patient&.telecom&.select { |t| t&.system == 'phone' }&.second&.value),
```

Comment:
Since the parsing/normalizing occurs here, we get potentially confusing error messages. If you include alphabetical characters in your phone number (e.g. `"(333) asdf"`), `Phonelib` will just remove them. This behavior results in an error message that indicates the normalized value, rather than the raw value (`"'+333' is not a valid phone number for 'Primary Telephone'"`).

The same concept would extend to any other fields for which we perform some kind of normalizing or transformation prior to updating the data model. A developer debugging could _probably_ figure out what's going on with the phone number example above, but I'm wondering if it could be more confusing in some cases. Can't think of any specific examples though really.

I also spent a while trying to think of how we could go about this so that we can capture the raw values in the error messages, and did not think the juice was worth the squeeze with any of the solutions I could come up with. Also tried googling a bit since this has got to be a relatively common concern for Rails API applications, but couldn't really find anything satisfactory. Let me know what you think Nick, and @kylietmo or @holmesie, whoever is reviewing here, would like to know your thoughts too. Is this fine as is? Is there a more elegant Rails-y solution here that I'm missing?

Reply:
Yeah this is a good question, and one I wasn't able to come up with an elegant solution for. I think this relates to a larger question of normalizing the fields. For example, I think long term it makes more sense to accept "Telephone call", "Telephone Call" and "telephone call", where right now we only accept "Telephone call".

Reply:
I think the more I think about this, the more confused I get. If we take the example of "Telephone call" discussed above, what is really bothering me is that normalization and validation are so similar to each other. To normalize we lowercase and remove spaces, and then map that lowercased, no-spaces version to the normalized string, in this case "Telephone call". So if we successfully normalize, we have essentially already done validation, we will either have "Telephone call" or `nil`.

For that reason it is starting to feel to me like we may as well just normalize and validate all in one. This is what was happening already in the `ImportController`, but the code still does need to be on `Patient` so that the `APIController` can also use it. The possibility I'm thinking of is what if we normalized the data in the custom validators? So the `PhoneValidator` would actually replace the current value of `Patient.primary_telephone` if it could successfully normalize that to a valid value.

On the one hand this solves some problems for us. Normalization/validation happen all at once and don't have to repeat similar code, we have access to the original value, so the issue @tstrass noted above would be resolved, and the controllers get simpler. But on the other hand, it feels sneaky to have validators actually modify the model they are being called on...that freaks me out. Let me know what you think @tstrass and others.

Reply:
I've been debating this for entirely too long. And again, I'm shocked by the fact that we seem to be pretty much the only people on the internet with this concern in Rails... Anyway, I agree I don't like validators having side effects. The alternatives I can think of would be to put normalization in the controller, in the model attribute setters, or in the `ActiveRecord` `before_validation` callback. I don't love any of these solutions, but I guess I would go with `before_validation`. How would you feel about that @ngfreiter? I realize this doesn't exactly solve your concern about normalization and validation doing very similar things, so I think the best we can do is just separate out any common functions into the `ValidationHelpers`, and that will probably also mean you may need to implement more of the validations as custom validators so that you can use those methods. That still may mean that most of those methods are being called twice (on normalization and again on validation), but I don't have an issue with that. It will be nice if validation can still run independently of normalization. Thoughts?

Reply:
Oh don't worry, I've been debating just as long. I've thought about `before_validation`, but the thing that I've gotten stuck on there is that different controllers seem to need different sorts of normalization. For example, we don't need to normalize an ethnicity coming in from the `APIController`, since we create the value from a code, so we can ensure it is a normalized value. The `ImportController` lets the value just be added as a string. For that reason I'm leaning towards keeping normalization in the controllers. But that's just a lean. I do agree that validation and normalization should stay separate.

Reply:
Right, ok. I agree with your reasoning then that normalization should be performed in the controllers. I do see now that that since we effectively 3 different mediums of input (web app, import, API), normalizing that input should be taken care of in each respective controller. I think those methods that perform the normalization could still be class methods on the model, but regardless, they would still be called from the controller prior to saving or validating data. And then also we could potentially just reuse those methods in custom validators. Out of scope of this task, though.

Reply:
So for this task - @tstrass are you okay with the slightly misleading error messages that will be sent out here for the time being or do you want to address this piece in this PR? I will say for this case in particular, we could avoid doing the parsing of the phone number until after do a validation check in the controller before we go to save (as you guys have proposed doing more broadly).

Reply:
Yeah, I'm okay with just having imperfect error messages for now.

Reply:
Would you mind making a ticket for this improvement in the future and putting it on the backlog?

Reply:
https://tracker.codev.mitre.org/browse/SARAALERT-965